internal/terraform/evaluate.go

```diff
@@ -338,193 +347,126 @@ func (d *evaluationStateData) GetModule(addr addrs.ModuleCall, rng tfdiags.Sourc
 	}
 	outputConfigs := moduleConfig.Module.Outputs
 
-	// Collect all the relevant outputs that current exist in the state.
-	// We know the instance path up to this point, and the child module name,
-	// so we only need to store these by instance key.
-	stateMap := map[addrs.InstanceKey]map[string]cty.Value{}
-	for _, elem := range d.Evaluator.NamedValues.GetOutputValuesForModuleCall(d.ModulePath, addr).Elems {
-		outputAddr := elem.Key
-		val := elem.Value
-
-		_, callInstance := outputAddr.Module.CallInstance()
-		instance, ok := stateMap[callInstance.Key]
-		if !ok {
-			instance = map[string]cty.Value{}
-			stateMap[callInstance.Key] = instance
-		}
-
-		instance[outputAddr.OutputValue.Name] = val
-	}
-
-	// Get all changes that reside for this module call within our path.
-	// The change contains the full addr, so we can key these with strings.
-	changesMap := map[addrs.InstanceKey]map[string]*plans.OutputChangeSrc{}
-	for _, change := range d.Evaluator.Changes.GetOutputChanges(d.ModulePath, addr) {
-		_, callInstance := change.Addr.Module.CallInstance()
-		instance, ok := changesMap[callInstance.Key]
-		if !ok {
-			instance = map[string]*plans.OutputChangeSrc{}
-			changesMap[callInstance.Key] = instance
+	// We don't do instance expansion during validation, and so we need to
+	// return an unknown value. Technically we should always return
+	// cty.DynamicVal here because the final value during plan will always
+	// be an object or tuple type with unpredictable attributes/elements,
+	// but because we never actually carry values forward from validation to
+	// planning we lie a little here and return unknown list and map types,
+	// just to give us more opportunities to catch author mistakes during
+	// validation.
+	//
+	// This means that in practice any expression that refers to a module
+	// call must be written to be valid for either a collection type or
+	// structural type of similar kind, so that it can be considered as
+	// valid during both the validate and plan walks.
+	if d.Operation == walkValidate {
+		atys := make(map[string]cty.Type, len(outputConfigs))
+		for name := range outputConfigs {
+			atys[name] = cty.DynamicPseudoType // output values are dynamically-typed
 		}
+		instTy := cty.Object(atys)
 
-		instance[change.Addr.OutputValue.Name] = change
+		switch {
+		case callConfig.Count != nil:
+			return cty.UnknownVal(cty.List(instTy)), diags
```

Comment:
What I wrote here is almost identical to some logic in the previous code, with the only difference being that it now does `return cty.UnknownVal(...)` instead of `ret = cty.UnknownVal`. Here's the relevant snippet of code currently on `main`:

https://github.com/hashicorp/terraform/blob/3b4964270f3d11ff73c95a9de1a4a5089e0ebe9a/internal/terraform/evaluate.go#L504-L525

I agree that this feels pretty suspect, but my goal here was to retain the previous behavior exactly, and reverting back to using `cty.DynamicVal` here would break our usual compatibility guidelines because it would cause something to become _less known_ in a future release.

This special case has been here since https://github.com/hashicorp/terraform/pull/24751, from 2020, and so I think it's now been in enough releases for us to both feel relatively confident that it's not hurting anything and that we are blocked from changing this behavior under the compatibility promises.

Reply:
Thanks, I missed that this was just transplanted code. It looks like we didn't roll that one out :( I'm pretty certain you could construct the same failure modes as you can for resources, but expanded modules with complex dynamic objects are far less common than with resources.

Reply:
Indeed... the way I made peace with this while working on this PR was summarized in the comment just above: as long as the expressions that refer to `module.whatever` only use language features that are valid for both collection and structural types this will work okay.

Since we always run validate as part of preparing the configuration for other operations, `terraform plan` effectively has an implicit `terraform validate` embedded in it and thus there should not be any situation where something would be accepted by plan but not accepted by validate.

I would object to adding this special case now because we could not be sure that all existing references to `module.whatever` are in contexts where both collection types and structural types are valid, but the fact that this code has been here for four years suggests that most existing modules in the world already meet this constraint and that any new modules written today would meet this constraint by necessity (or else the module wouldn't actually work).